recipes-bsp/secure-boot-otp-provisioning/README.md

```diff
@@ -0,0 +1,154 @@
+# optional the third key
+#openssl req -x509 -newkey rsa:4096 -keyout custBmpk.pem -nodes -outform pem -out custBmpk.crt -sha256
+```
+
+> :warning:
+> Do remove the example dummy keys under the keys folder!
+
+## OTP programming
+
+The OTP programming data is generated by running the script `files/make-otpcmd.sh`.
+The generated `otpcmd.bin` should be programmed into the flash, and during the
+next booting, the first stage bootloader (SEBoot) will detect it and will try to
+program it into the OTP area.
+
+To ease the key provisioning tasks, we provided an example firmware image in
+this repo to build an image containing both:
+
+- the otpcmd.bin
+- signed firmware
+
+so that the first booting of signed firmware and OTP programming could be combined
+together.
+
+You can make your own implement based on this blueprint, or you can use the script
+directly.
+
+### Building the signed firmware image containing the key provisioning data
+
+The next step is to build the signed firmware image together with the key
+provisioning data.
+
+Before start the build, make sure the currently using key set is copied to
+`recipes-bsp/u-boot/files/keys/custMpk.pem`.
+
+```shell
+./kas-container build kas-iot2050-boot-pg2.yml:kas/opt/secure-boot.yml:kas/opt/otpcmd/key-provision.yml
+```
+
+> Warning: The default key sets within kas/opt/key-provision.yml only contain
+> MPK and SMPK. To include BMPK as well, please refer to
+[Flexible key provisioning](#flexible-key-provisioning).
+
+### Update the firmware to provisioning the keys
+
+Copy the built firmware image to IOT2050 and run the firmware update against the
+new firmware image:
+
+```shell
+iot2050-firmware-update -f iot2050-pg2-image-boot.bin
+```
+
+Reboot the IOT2050 and follow the instruction on the device. The debug serial
+port has to be connected to give your consent to the provisioning routine.
+
+> Warning: after provisioning, all application must be signed with the currently
+> using key.
+
+The key provisioning just need be run once. Later secure boot firmware building
+should drop the `kas/opt/key-provision.yml`
+
+### Revoke the leaked currently using key
+
+If unfortunately the currently using key is leaked, you could switch to use the
+next keysets as the effective one.
+
+First make sure the new key is copied to `recipes-bsp/u-boot/files/keys/custMpk.pem`.
+
+Then start to building the new key signed firmware together with the key switching
+otpcmd data.
+
+```shell
+./kas-container build kas-iot2050-boot-pg2.yml:kas/opt/secure-boot.yml:kas/opt/otpcmd/key-switch.yml
+```
+
+> Warning: The default switching within kas/opt/key-switch.yml is from MPK to SMPK.
+> If you want to switch from SMPK to BMPK, please refer to
+[Flexible key provisioning](#flexible-key-provisioning).
+
+Then follow the exact same steps as defined in above, to update the firmware and
+follow the instructions on device after rebooting.
```

Comment:
Does the use need to confirm key rotation interactively as well? May create a tricky in-field maintenance story.

Reply:
Dup, close.